<h1>Perceptron</h1>

O Perceptron é um modelo clássico de machine learning baseado em redes neurais, introduzido por Frank Rosenblatt em 1958. É considerado a unidade básica de uma rede neural.

<h3>🎯 O que é e qual o objetivo?</h3>
O Perceptron é um modelo linear de classificação binária. Ele tenta encontrar um hiperplano que separa os dados em duas classes, baseado em uma combinação linear das entradas com pesos ajustáveis.


<h4>Y^= sign (𝑤⋅𝑥+𝑏)</h4>

<ul>
    <li>x: vetor de entrada</li>
    <li>w: pesos</li>
    <li>b: viés</li>
    <li>sign: função que retorna +1 ou -1 (ou 0/1)</li>
</ul>

<strong>Objetivo:</strong> aprender os pesos w que permitem classificar corretamente os dados de entrada em uma das classes.

<h3>✅ Pré-requisitos</h3>
<ol>
    <li>Problema de classificação binária (duas classes).</li>
    <li>Dados linearmente separáveis (ou quase).</li>
    <li>Dados numéricos ou convertidos para numéricos.</li>
</ol>

<h3>✅ Vantagens</h3>
<ul>
    <li>Simples e rápido de implementar e treinar.</li>
    <li>Base para redes neurais mais complexas.</li>
    <li>Útil como modelo introdutório para aprendizado supervisionado.</li>
</ul>

<h3>❌ Desvantagens</h3>
<ul>
    <li>Funciona apenas com problemas linearmente separáveis.</li>
    <li>Não converge se os dados não puderem ser perfeitamente separados por uma linha/hiperplano.</li>
    <li>Desempenho limitado comparado a modelos modernos como SVM, Random Forest, etc.</li>
</ul>

<h3>💼 Aplicações comuns</h3>
Apesar de hoje ser mais um modelo didático, o Perceptron tem aplicações em:
<ul>
    <li>Classificadores binários simples (por exemplo, detectar se um número é positivo ou negativo).</li>
    <li>Base para redes neurais profundas.</li>
    <li>Problemas educacionais e teóricos.</li>
    <li>Reconhecimento de padrões básicos (ex: dígitos manuscritos com binarização).</li>
</ul>













In [2]:

import numpy as np
import pandas as pd
import sklearn

from pandas import Series, DataFrame
from sklearn import datasets
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report


<h3>Importanto o perceptron</h3>

In [3]:
from sklearn.linear_model import Perceptron


<h2>Carregando o dataset</h2>

In [5]:
df = pd.read_csv('iris.data.csv', header=None)
df.columns = ['Sepal Heigth', 'Sepal Width', 'Petal Heigth', 'Petal Width', 'Spec']
iris_data = df[['Sepal Heigth', 'Sepal Width', 'Petal Heigth', 'Petal Width']]
X = iris_data.values
iris_target = df['Spec']
y = iris_target.values
print(f'Iris Data Sample:\n{X[0:5]}\n')
print(f'Iris Target Sample:\n{y[0:5]}\n')


Iris Data Sample:
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]

Iris Target Sample:
['setosa' 'setosa' 'setosa' 'setosa' 'setosa']



<h2>Separando os dados em dados de treinamento e dados de teste</h2>

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

<h3>Normalizando os dados</h3>

In [ ]:
standardize = StandardScaler()
standardised_X_test = standardize.fit_transform(X_test)
standardised_X_train = standardize.fit_transform(X_train)

In [7]:
standardised_X_test[0:5]

array([[-0.79576404,  1.10573515, -1.36131706, -1.41523807],
       [ 2.2320211 , -0.81728251,  1.77571514,  1.53707976],
       [-0.79576404,  1.34611236, -1.41941025, -1.27465151],
       [-1.49448369, -1.53841413, -1.47750344, -1.27465151],
       [ 0.3687687 , -0.33652809,  0.49766499, -0.00937244]])

<h2>Criando o perceptron </h2>

In [ ]:
perceptron = Perceptron(max_iter=50, eta0=0.15, tol=1e-3, random_state=15)
perceptron.fit(standardised_X_train, y_train.ravel())

Perceptron(eta0=0.15, max_iter=50, random_state=15)

<h2>Fazendo um previsão </h2>

In [ ]:
y_pred = perceptron.predict(standardised_X_test)

print(f'Dados de teste: {y_test}\n')
print(f'Dados previstos: {y_pred}\n')

Dados de teste: ['setosa' 'virginica' 'setosa' 'setosa' 'versicolor' 'virginica'
 'versicolor' 'setosa' 'setosa' 'setosa' 'virginica' 'versicolor'
 'versicolor' 'versicolor' 'virginica' 'virginica' 'versicolor'
 'versicolor' 'setosa' 'virginica' 'setosa' 'versicolor' 'virginica'
 'versicolor' 'setosa' 'versicolor' 'versicolor' 'versicolor' 'virginica'
 'virginica']

Dados previstos: ['setosa' 'virginica' 'setosa' 'setosa' 'versicolor' 'virginica'
 'virginica' 'setosa' 'setosa' 'setosa' 'virginica' 'virginica'
 'virginica' 'versicolor' 'virginica' 'virginica' 'versicolor'
 'versicolor' 'setosa' 'virginica' 'setosa' 'versicolor' 'virginica'
 'versicolor' 'setosa' 'virginica' 'versicolor' 'versicolor' 'virginica'
 'virginica']



<h2>Relatório de Classificação</h2>

In [31]:
print('Relatório da classificação\n\n',classification_report(y_test,y_pred))

Relatório da classificação

               precision    recall  f1-score   support

      setosa       1.00      1.00      1.00         9
  versicolor       1.00      0.67      0.80        12
   virginica       0.69      1.00      0.82         9

    accuracy                           0.87        30
   macro avg       0.90      0.89      0.87        30
weighted avg       0.91      0.87      0.87        30



<h3>Interpretação</h3>
<h4>Por classe</h4>
<h5>Setosa</h5>
<ul><li>Precision = 1.00: tudo que foi classificado como setosa estava correto.</li>
    <li>Recall = 1.00: o modelo identificou todas as setosa corretamente.</li>
    <li>F1-score = 1.00: desempenho perfeito — sem falsos positivos nem falsos negativos.</li>
    <li>Support = 9: havia 9 amostras verdadeiras de setosa.</li>
</ul>
<h5>Versicolor</h5>
<ul><li>Precision = 1.00: todas as predições como versicolor estavam corretas.</li>
    <li>Recall = 0.67: o modelo acertou apenas 67% das versicolor reais.</li>
    <li>F1-score = 0.80: penaliza a baixa cobertura (recall).</li>
    <li>Support = 12: 12 amostras reais de versicolor.</li>
</ul>
<h5>Virginica</h5>
<ul><li>Precision = 0.69: 69% dos exemplos classificados como virginica estavam corretos.</li>
    <li>Recall = 1.00: o modelo identificou todos os virginica reais, mas com falsos positivos.</li>
    <li>F1-score = 0.82: valor intermediário — bom recall, precisão moderada.</li>
    <li>Support = 9</li>
</ul>
<h4>Métricas globais</h4>
<ul><li>Accuracy: 87% das predições foram corretas no total.</li>
    <li>Macro avg</li>
    <ul><li>Média simples entre as três classes (sem pesar pelo número de amostras). Útil quando as classes têm peso semelhante.</li></ul>
    <li>Weighted avg</li>
    <ul><li>Média ponderada de cada classe, de acordo com seu support.</li></ul>
</ul>

<h4> Conclusões</h4>
O modelo classificou perfeitamente a classe setosa.
Erra parte das versicolor, mesmo que tenha alta precisão (possivelmente devido a pouca amostra ou confusão com virginica).
Em virginica, o modelo encontra todos corretamente, mas comete falsos positivos (baixa precisão).